### ISSUE TITLE
Guide/Doc for correlation

### ISSUE THREAD

**Issue by leonwangcn** [sipcapture/homer#343]
Hello,

Thanks for your good work on HOMER 7.
Regarding protocols correlation, do you have some examples or Doc available to let me know how to do it? We are trying to develop H248/MEGACO protocol support in HOMER 7 and then make it be correlated with SIP...

Thanks,
Leon

**Comment by lmangani** [comment]
Hello,

Thanks for your interest and for your development efforts! As long as your protocol produces a JSON object into H7, the correlation can be abstracted using the "Mapping" settings which control cross-field correlation between protocols by field and range. We'll add some new Docs to the Wiki in the coming weeks, and of course we'll be happy to help you achieve that and prepare the initials once you have some objects formed in DB for sure!

**Comment by leonwangcn** [comment]
Thanks a lot!
Regarding JSON object, would you please help say more? I am trying to see how to implement the correlation step by step, since I have succeeded in sending SIP and MEGACO msgs into database and decoded in HOMER 7. I tried to modify the mappings (look_up id, range) but failed to correlate them...

br,
Leon

**Comment by lmangani** [comment]
I'd suggest opening a PR with your changes and/or detailing your work so far and I'm sure our team will be able to fill in the gaps.

**Comment by leonwangcn** [comment]
How to open a PR? The other point, do you have any example to show how to do correlation, which could be a reference?
Thanks!

br,
Leon

**Comment by lmangani** [comment]
Hi @leonwangcn inside HOMER 7 settings under the `Mapping` tab you will find the current correlation - this should be a good start while we prepare a proper guide, feel free to ask anything here.

**Comment by leonwangcn** [comment]
Thanks! 
Just now I watched Kamailio world 2019 Homer7 demo, in which it correlated LOG and SIP, under HOMER7+HEPGEN ENV.  I did not see they did some configurations in Mapping tab. Does it mean the default configuration support LOG + SIP correlation?

**Comment by leonwangcn** [comment]
To support SIP correlation with Megaco, in profile of SIP, I changed the lookup_id to be 2 (id for megaco) but still no output...should I change more such as lookup_field?

 "source_field": "data_header.callid",
    "lookup_id": 2,// id for megaco is 2
    "lookup_profile": "default",
    "lookup_field": "sid",
    "lookup_range": [
      -300,
      200

Thanks,
Leon

**Comment by adubovikov** [comment]
@leonwangcn just before we start make a mapping, do you have already data in your hep_proto_2_default or hep_proto_2_call  table ? Did you send the MEGACO as a JSON payload with decoded fields. How did you set the correlation_id to MEGACO ?  Did you make an external lookup to correlate SDP from SIP calls with MEGACO media_ip ? 

please provide all information and we can help you. The approach is very interesting, specially in case that we already support ISUP/SIPGTRAN (please check the captagent/module/ss7)  and add MEGACO or MGCP will close the circle of FULL correlation.

Looking forward to get all information! Thank you!

**Comment by leonwangcn** [comment]
Hello Alexandr,

Thank you very much for your quick feedback! And glad to receive from you as I just watched your presentation/demo :).

In HOMER 7, by referring to SIP parsing implementation, we are trying to provide MEGACO decoding/filtering support...
Now, we are doing unit test after draft work done. We used simulators to make a call through which to generate SIP and MEGACO msgs. With HEP encapsulation (only hep id were set for them), we sent the SIP and MEGACO msgs to HOMER7 and wanted to see the correlation...
Per your questions, it seems we need to do much work in HOMER7 side besides changes in mapping tab. Would you please help point more specifically about what kind of work we need to do? Thank you very much!

Btw, it seems Homer7 is not much stable yet...for example, after we search SIP msgs with a filter, if we remove the filter and search ag, there will be nothing showed...seems the filter was written uncorrectly...maybe just set it as {"hepid":1,"profile":"call"} is ok...

br,
Leon

**Comment by lmangani** [comment]
@leonwangcn push your code to a repository so we can take a look and help you complete it. Alternatively provide a PCAP with the HEP you're sending and/or a raw version of it for comments and suggestions.

**Comment by adubovikov** [comment]
@leonwangcn about filter, please open an another issue and describe how to reproduce it and we will ASAP check it! 
About MEGACO, unfortunately I didn't receive any answer here. Did you store MEGACO as an JSON object already in the hep_proto_2_xxx or not ? How did you send it, because as I said already in the captagent we have only SIGTRAN but not MGCP/MEGACO. As Lorenzo asked already, please provide all details and pcaps if it's possible to be in sync, because currently I am a bit lost what and how it has been implemented from your side :-) Thank you!

**Comment by leonwangcn** [comment]
[hep.txt](https://github.com/sipcapture/homer/files/3186324/hep.txt)
Please change the expansion name to be .pcap
Here is the HEP msgs we generated out...

**Comment by adubovikov** [comment]
so, first, you don't send it as a real JSON format, this means somebody should make preparation to store in into DB as a property JSON format. Second you don't send any correlation id in the HEP header. How Homer should correlate it ? ;-) We don't do magic tricks :-). For now, I suggest you send the MEGACO as a log HEP type (100) and set somehow a correlation_id that will point to SIP session.

but again, it looks very promising!
![Screenshot from 2019-05-16 12-51-59](https://user-images.githubusercontent.com/4513061/57848771-f5fb3380-77d9-11e9-9ec5-802f59d7a144.png)
!

**Comment by lmangani** [comment]
@leonwangcn I fully agree with @adubovikov set to HEP Type 100 and set the `correlation_id` field to the Call-ID of the corresponding session. What is the sender of HEP is? Is it Open-Source?

**Comment by leonwangcn** [comment]
Hello Lorenzo, Alexandr,
Thanks a lot!
I will do a try tomorrow according to what you said.
Before it, I need your help to clarify one more question.
In Mapping tab, we may see some default mapping rules in Homer7.
For example, MEGACO, with HEP ID as 2.  
In HEP3, protocol type for MEGACO is 7, why did you set HEP ID 2 here?

br,
Leon

**Comment by lmangani** [comment]
@adubovikov As far as I know there's no default for MEGACO in Mappings yet, with neither ID. 
@leonwangcn Is this perhaps the one you've added?

**Comment by adubovikov** [comment]
no clue, by default we dont have it:
https://github.com/sipcapture/homer-app/blob/master/seeds/02_Mapping.js

maybe I missed something and somebody added it, but in the master I don't see that

**Comment by leonwangcn** [comment]
Close this issue since I have got almost clear...thanks a ton!
